src/csharp/Microsoft.Spark/Broadcast.cs

```diff
@@ -170,21 +171,65 @@ private JvmObjectReference CreateBroadcast(SparkContext sc, T value)
             bool encryptionEnabled = bool.Parse(
                 sc.GetConf().Get("spark.io.encryption.enabled", "false"));
 
+            var _pythonBroadcast = (JvmObjectReference)javaSparkContext.Jvm.CallStaticJavaMethod(
+                "org.apache.spark.api.python.PythonRDD",
+                "setupBroadcast",
+                _path);
+
             if (encryptionEnabled)
             {
-                throw new NotImplementedException("Broadcast encryption is not supported yet.");
+                var pair = (JvmObjectReference[])_pythonBroadcast.Invoke("setupEncryptionServer");
+
+                using ISocketWrapper socket = SocketFactory.CreateSocket();
+                socket.Connect(
+                    IPAddress.Loopback,
+                    (int)pair[0].Invoke("intValue"),
+                    (string)pair[1].Invoke("toString"));
+                WriteInChunks(value, 8192, socket.OutputStream);
+                socket.OutputStream.Close();
+                _pythonBroadcast.Invoke("waitTillDataReceived");
             }
             else
             {
                 WriteToFile(value);
             }
 
-            var pythonBroadcast = (JvmObjectReference)javaSparkContext.Jvm.CallStaticJavaMethod(
-                "org.apache.spark.api.python.PythonRDD",
-                "setupBroadcast",
-                _path);
+            return (JvmObjectReference)javaSparkContext.Invoke("broadcast", _pythonBroadcast);
+        }
 
-            return (JvmObjectReference)javaSparkContext.Invoke("broadcast", pythonBroadcast);
+        /// TODO: This is not performant as it writes to stream only after serializing the whole
+        /// value, instead of serializing and writing in chunks like Python.
+        /// <summary>
+        /// Function to write the broadcast value into the stream in fixed size chunks.
+        /// </summary>
+        /// <param name="value">Broadcast value to be written to the stream</param>
+        /// <param name="chunkSize">Size of chunk to write values in</param>
+        /// <param name="stream">Stream connecting to encryption server to write value to</param>
+        private void WriteInChunks(object value, int chunkSize, Stream stream)
+        {
+            var formatter = new BinaryFormatter();
+            using var ms = new MemoryStream();
+            formatter.Serialize(ms, value);
+            byte[] valueBytes = ms.ToArray();
+            int bytesRemaining = valueBytes.Length;
+            int bytePos = 0;
+            while (bytesRemaining > 0)
+            {
+                if (bytesRemaining < chunkSize)
+                {
+                    SerDe.Write(stream, bytesRemaining);
+                    SerDe.Write(stream, valueBytes, bytePos, bytesRemaining);
+                    bytesRemaining = 0;
+                }
+                else
+                {
+                    SerDe.Write(stream, chunkSize);
```

Comment:
can you add a test where serialized object is bigger than the chunk size?

Reply:
wait, why do you need to do this chunking instead of writing valueBytes at once? It's not that stream is flushing after each chunk. I don't get it. The Scala side will still wait for all bytes to be received, no?

Reply:
Btw, can you create a big broadcast object (>100MB) and stress test it once?

Reply:
Scala just expects a chunk size (number of bytes to expect) first and then the value bytes. I tested by sending all bytes representation of the broadcast value prepended with the size of byte array and then appending with -1 and that works. I just added the chunking logic here to mimic python sending 8192 sized chunks at a time, but since we are not doing the serializing accordingly I think we can just write all bytes at once. I will stress test it with a large broadcast value to verify.

Reply:
Added a stress test with large broadcast value with the current logic of sending all bytes at once and it works. Please let me know if this looks ok, thanks!